In [1]:
import pandas as pd

<div><b>Air quality Nitrate data</b></div>
<div>
<p></p>
<p>For this tutorial, air quality data about $NO_2$ is used, made available by <a href="https://openaq.org">OpenAQ</a> and downloaded using the <a href="http://dhhagan.github.io/py-openaq/index.html">py-openaq</a> package.</p>

<p>The <code>air_quality_no2_long.csv</code> data set provides $NO_2$ values for the measurement stations <em>FR04014</em>, <em>BETR801</em> and <em>London Westminster</em> in respectively Paris, Antwerp and London.</p>
<p></p>
<a href="https://github.com/pandas-dev/pandas/tree/main/doc/data/air_quality_no2_long.csv">To raw data</a>
</div>

In [2]:
!curl https://raw.githubusercontent.com/pandas-dev/pandas/refs/heads/main/doc/data/air_quality_no2_long.csv -o air_quality_no2_long.csv

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  132k  100  132k    0     0   499k      0 --:--:-- --:--:-- --:--:--  501k


In [4]:
air_quality_no2 = pd.read_csv("air_quality_no2_long.csv", parse_dates=True)

In [5]:
air_quality_no2 = air_quality_no2[["date.utc", "location", "parameter", "value"]]

In [6]:
air_quality_no2.head()

,date.utc,location,parameter,value
0,2019-06-21 00:00:00+00:00,FR04014,no2,20.0
1,2019-06-20 23:00:00+00:00,FR04014,no2,21.8
2,2019-06-20 22:00:00+00:00,FR04014,no2,26.5
3,2019-06-20 21:00:00+00:00,FR04014,no2,24.9
4,2019-06-20 20:00:00+00:00,FR04014,no2,21.4


<div><b>Air quality Particulate matter data</b></div>
<div>
<p></p>
<p>For this tutorial, air quality data about Particulate matter less than 2.5 micrometers is used, made available by <a href="https://openaq.org">OpenAQ</a> and downloaded using the <a href="http://dhhagan.github.io/py-openaq/index.html">py-openaq</a> package.</p>

<p>The <code class="docutils literal notranslate">air_quality_pm25_long.csv</code> data set provides $PM_{25}$ values for the measurement stations <em>FR04014</em>, <em>BETR801</em> and <em>London Westminster</em> in respectively Paris, Antwerp and London.</p>
<p></p>
<a href="https://github.com/pandas-dev/pandas/tree/main/doc/data/air_quality_pm25_long.csv">To raw data</a>
</div>

In [7]:
!curl https://raw.githubusercontent.com/pandas-dev/pandas/refs/heads/main/doc/data/air_quality_pm25_long.csv -o air_quality_pm25_long.csv

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 79525  100 79525    0     0   353k      0 --:--:-- --:--:-- --:--:--  354k


In [8]:
air_quality_pm25 = pd.read_csv("air_quality_pm25_long.csv", parse_dates=True)

In [9]:
air_quality_pm25 = air_quality_pm25[["date.utc", "location", "parameter", "value"]]

In [10]:
air_quality_pm25.head()

,date.utc,location,parameter,value
0,2019-06-18 06:00:00+00:00,BETR801,pm25,18.0
1,2019-06-17 08:00:00+00:00,BETR801,pm25,6.5
2,2019-06-17 07:00:00+00:00,BETR801,pm25,18.5
3,2019-06-17 06:00:00+00:00,BETR801,pm25,16.0
4,2019-06-17 05:00:00+00:00,BETR801,pm25,7.5


<h1>How to combine data from multiple tables</h1>

# <h2>Concatenating objects</h2>

<img class="align-center" src="https://pandas.pydata.org/docs/_images/08_concat_row.svg">

### <h4>I want to combine the measurements of $NO_2$ and $PM_{25}$, two tables with a similar structure, in a single table.</h4>

In [11]:
air_quality = pd.concat([air_quality_pm25, air_quality_no2], axis=0)

In [12]:
air_quality.head()

,date.utc,location,parameter,value
0,2019-06-18 06:00:00+00:00,BETR801,pm25,18.0
1,2019-06-17 08:00:00+00:00,BETR801,pm25,6.5
2,2019-06-17 07:00:00+00:00,BETR801,pm25,18.5
3,2019-06-17 06:00:00+00:00,BETR801,pm25,16.0
4,2019-06-17 05:00:00+00:00,BETR801,pm25,7.5


<p>The <a href="https://pandas.pydata.org/docs/reference/api/pandas.concat.html#pandas.concat" title="pandas.concat"><code>concat()</code></a> function performs concatenation operations of multiple
tables along one of the axes (row-wise or column-wise).</p>

<p>By default concatenation is along axis 0, so the resulting table combines the rows of the input tables. Let’s check the shape of the original and the concatenated tables to verify the operation:</p>

In [15]:
print('Shape of the air_quality_pm25 table: ', air_quality_pm25.shape)

Shape of the air_quality_pm25 table:  (1110, 4)


In [16]:
print('Shape of the air_quality_no2 table: ', air_quality_no2.shape)

Shape of the air_quality_no2 table:  (2068, 4)


In [17]:
print('Shape of the resulting air_quality table: ', air_quality.shape)

Shape of the resulting air_quality table:  (3178, 4)


<p>Hence, the resulting table has 3178 = 1110 + 2068 rows.</p>

<div class="alert alert-block alert-info">
<p>Note</p>
<p>The <strong>axis</strong> argument will return in a number of pandas
methods that can be applied <strong>along an axis</strong>.
A <code>DataFrame</code> has two corresponding axes: the first running vertically downwards across rows (axis 0), and the second running horizontally across columns (axis 1).
Most operations like concatenation or summary statistics are by default
across rows (axis 0), but can be applied across columns as well.</p>
</div>

<p>Sorting the table on the datetime information illustrates also the
combination of both tables, with the <code>parameter</code> column defining the origin of the table (either <code>no2</code> from table <code>air_quality_no2</code> or <code>pm25</code> from table <code>air_quality_pm25</code>):</p>

In [18]:
air_quality = air_quality.sort_values("date.utc")

In [19]:
air_quality.head()

,date.utc,location,parameter,value
2067,2019-05-07 01:00:00+00:00,London Westminster,no2,23.0
1003,2019-05-07 01:00:00+00:00,FR04014,no2,25.0
100,2019-05-07 01:00:00+00:00,BETR801,pm25,12.5
1098,2019-05-07 01:00:00+00:00,BETR801,no2,50.5
1109,2019-05-07 01:00:00+00:00,London Westminster,pm25,8.0


<p>In this specific example, the <code>parameter</code> column provided by the data ensures that each of the original tables can be identified. This is not always the case. The <code>concat</code> function provides a convenient solution with the <code>keys</code> argument, adding an additional (hierarchical) row index. For example:</p>

In [20]:
air_quality_ = pd.concat([air_quality_pm25, air_quality_no2], keys=["PM25", "NO2"])

In [21]:
air_quality_.head()

date.utc location parameter  value
PM25 0  2019-06-18 06:00:00+00:00  BETR801      pm25   18.0
     1  2019-06-17 08:00:00+00:00  BETR801      pm25    6.5
     2  2019-06-17 07:00:00+00:00  BETR801      pm25   18.5
     3  2019-06-17 06:00:00+00:00  BETR801      pm25   16.0
     4  2019-06-17 05:00:00+00:00  BETR801      pm25    7.5

<div class="alert alert-block alert-info">
<p>Note</p>
<p>The existence of multiple row/column indices at the same time
has not been mentioned within these tutorials.
<em>Hierarchical indexing</em> or <em>MultiIndex</em> is an advanced and powerful pandas feature to analyze higher dimensional data.</p>
<p>Multi-indexing is out of scope for this pandas introduction.
For the moment, remember that the function <code>reset_index</code> can be used to convert any level of an index to a column, e.g.
<code>air_quality.reset_index(level=0)</code></p>
</div>

<div>
<p><b>To user guide</b> Feel free to dive into the world of multi-indexing at the user guide section on <a href="https://pandas.pydata.org/docs/user_guide/advanced.html#advanced">advanced indexing</a>.</p>
</div>

<div>
<p><b>To user guide</b> More options on table concatenation (row and column wise) and how <code>concat</code> can be used to define the logic (union or intersection) of the indexes on the other axes is provided at the section on <a href="https://pandas.pydata.org/docs/user_guide/merging.html#merging-concat">object concatenation</a>.</p>
</div>

# <h2>Join tables using a common identifier</h2>

<img class="align-center" src="https://pandas.pydata.org/docs/_images/08_merge_left.svg">

### <h4>Add the station coordinates, provided by the stations metadata table, to the corresponding rows in the measurements table.</h4>

<div class="alert alert-block alert-danger">
<p>Warning</p>
<p>The air quality measurement station coordinates are stored in a data
file <code>air_quality_stations.csv</code>, downloaded using the
<a href="http://dhhagan.github.io/py-openaq/index.html">py-openaq</a> package.</p>
</div>

In [22]:
!curl https://raw.githubusercontent.com/pandas-dev/pandas/refs/heads/main/doc/data/air_quality_stations.csv -o air_quality_stations.csv

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1923  100  1923    0     0  14753      0 --:--:-- --:--:-- --:--:-- 14792


In [23]:
stations_coord = pd.read_csv("air_quality_stations.csv")

In [24]:
stations_coord.head()

,location,coordinates.latitude,coordinates.longitude
0,BELAL01,51.23619,4.38522
1,BELHB23,51.17030,4.34100
2,BELLD01,51.10998,5.00486
3,BELLD02,51.12038,5.02155
4,BELR833,51.32766,4.36226


<div class="alert alert-block alert-info">
<p>Note</p>
<p>The stations used in this example (FR04014, BETR801 and London
Westminster) are just three entries enlisted in the metadata table.
We only want to add the coordinates of these three to the measurements
table, each on the corresponding rows of the <code>air_quality</code> table.</p>
</div>

In [25]:
air_quality.head()

,date.utc,location,parameter,value
2067,2019-05-07 01:00:00+00:00,London Westminster,no2,23.0
1003,2019-05-07 01:00:00+00:00,FR04014,no2,25.0
100,2019-05-07 01:00:00+00:00,BETR801,pm25,12.5
1098,2019-05-07 01:00:00+00:00,BETR801,no2,50.5
1109,2019-05-07 01:00:00+00:00,London Westminster,pm25,8.0


In [26]:
air_quality = pd.merge(air_quality, stations_coord, how="left", on="location")

In [27]:
air_quality.head()

,date.utc,location,parameter,value,coordinates.latitude,coordinates.longitude
0,2019-05-07 01:00:00+00:00,London Westminster,no2,23.0,51.49467,-0.13193
1,2019-05-07 01:00:00+00:00,FR04014,no2,25.0,48.83724,2.39390
2,2019-05-07 01:00:00+00:00,FR04014,no2,25.0,48.83722,2.39390
3,2019-05-07 01:00:00+00:00,BETR801,pm25,12.5,51.20966,4.43182
4,2019-05-07 01:00:00+00:00,BETR801,no2,50.5,51.20966,4.43182


<p>Using the <a href="https://pandas.pydata.org/docs/reference/api/pandas.merge.html#pandas.merge" title="pandas.merge"><code>merge()</code></a> function, for each of the rows in the <code>air_quality</code> table, the corresponding coordinates are added from the <code>air_quality_stations_coord</code> table.
Both tables have the column <code>location</code> in common which is used as a key to combine the information.
By choosing the <code>left</code> join, only the locations available in the <code>air_quality</code> (left) table, i.e.&nbsp;FR04014, BETR801 and London Westminster, end up in the resulting table.
The <code>merge</code> function supports multiple join options similar to database-style operations.</p>

### <h4>Add the parameters’ full description and name, provided by the parameters metadata table, to the measurements table.</h4>

<div class="alert alert-block alert-danger">
<p>Warning</p>
<p>The air quality parameters metadata are stored in a data file
<code>air_quality_parameters.csv</code>, downloaded using the
<a href="http://dhhagan.github.io/py-openaq/index.html">py-openaq</a> package.</p>
</div>

In [52]:
!curl https://raw.githubusercontent.com/pandas-dev/pandas/8cf2e43e1ca2a50f010fd37da5a01e8a3a6f1822/doc/data/air_quality_parameters.csv -o air_quality_parameters.csv

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   255  100   255    0     0   1483      0 --:--:-- --:--:-- --:--:--  1491


In [53]:
air_quality_parameters = pd.read_csv("air_quality_parameters.csv")

In [54]:
air_quality_parameters.head()

,id,description,name
0,bc,Black Carbon,BC
1,co,Carbon Monoxide,CO
2,no2,Nitrogen Dioxide,NO2
3,o3,Ozone,O3
4,pm10,Particulate matter less than 10 micrometers in...,PM10


In [55]:
air_quality = pd.merge(air_quality, air_quality_parameters, how='left', left_on='parameter', right_on='id')

In [56]:
air_quality.head()

,date.utc,location,parameter,value,coordinates.latitude,coordinates.longitude,id,description,name
0,2019-05-07 01:00:00+00:00,London Westminster,no2,23.0,51.49467,-0.13193,no2,Nitrogen Dioxide,NO2
1,2019-05-07 01:00:00+00:00,FR04014,no2,25.0,48.83724,2.39390,no2,Nitrogen Dioxide,NO2
2,2019-05-07 01:00:00+00:00,FR04014,no2,25.0,48.83722,2.39390,no2,Nitrogen Dioxide,NO2
3,2019-05-07 01:00:00+00:00,BETR801,pm25,12.5,51.20966,4.43182,pm25,Particulate matter less than 2.5 micrometers i...,PM2.5
4,2019-05-07 01:00:00+00:00,BETR801,no2,50.5,51.20966,4.43182,no2,Nitrogen Dioxide,NO2


<p>Compared to the previous example, there is no common column name.
However, the <code>parameter</code> column in the <code>air_quality</code> table and the <code>id</code> column in the <code>air_quality_parameters_name</code> both provide the measured variable in a common format.
The <code>left_on</code> and <code>right_on</code> arguments are used here (instead of just <code>on</code>) to make the link between the two tables.</p>

<div>
<p><b>To user guide</b> Pandas supports also inner, outer, and right joins. More information on join/merge of tables is provided in the user guide section on <a href="https://pandas.pydata.org/docs/user_guide/merging.html#merging-join">database style merging of tables</a>.
Or have a look at the <a href="https://pandas.pydata.org/docs/getting_started/comparison/comparison_with_sql.html#compare-with-sql-join">comparison with SQL</a> page.</p>
</div>

<div>
<h4>REMEMBER</h4><ul>
<li><p>Multiple tables can be concatenated both column-wise and row-wise using the <code>concat</code> function.</p></li>
<li><p>For database-like merging/joining of tables, use the <code>merge</code> function.</p></li>
</ul>
</div>

<div>
<p><b>To user guide</b> See the user guide for a full description of the various <a href="https://pandas.pydata.org/docs/user_guide/merging.html#merging">facilities to combine data tables</a>.</p>
</div>